independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/ContextDataMap.java

```diff
@@ -0,0 +1,119 @@
+    ContextDataMap(Set<Annotation> interceptorBindings) {
+        this.interceptorBindings = Objects.requireNonNull(interceptorBindings);
+    }
+
+    private Map<String, Object> getDelegateForRead() {
+        if (delegate == null) {
+            return Collections.EMPTY_MAP;
+        } else {
+            return delegate;
+        }
+    }
+
+    private Map<String, Object> getDelegateForWrite(final int sizeHint) {
+        if (delegate == null) {
+            this.delegate = new HashMap<>(sizeHint, 1.0f);
+        }
+        return delegate;
+    }
+
+    // ** Implement all methods from Map interface **
+
+    @Override
+    public void clear() {
+        throw new UnsupportedOperationException("Not allowed to clear the context data map");
+    }
+
+    @Override
+    public boolean containsKey(Object key) {
+        if (ArcInvocationContext.KEY_INTERCEPTOR_BINDINGS.equals(key)) {
+            return true;
+        }
+        return getDelegateForRead().containsKey(key);
+    }
+
+    @Override
+    public boolean containsValue(Object value) {
+        if (interceptorBindings.equals(value)) {
+            return true;
+        }
+        return getDelegateForRead().containsValue(value);
+    }
+
+    @Override
+    public Set entrySet() {
+        throw new UnsupportedOperationException("Not allowed to iterate the full content of the context data map");
+    }
+
+    @Override
+    public Object get(Object key) {
+        if (ArcInvocationContext.KEY_INTERCEPTOR_BINDINGS.equals(key)) {
+            return interceptorBindings;
+        }
+        return getDelegateForRead().get(key);
+    }
+
+    @Override
+    public boolean isEmpty() {
+        return false;
+    }
+
+    @Override
+    public Set keySet() {
+        throw new UnsupportedOperationException("Not allowed to iterate the full content of the context data map");
+    }
+
+    @Override
+    public void putAll(Map m) {
+        if (m.containsKey(ArcInvocationContext.KEY_INTERCEPTOR_BINDINGS)) {
+            throw new IllegalArgumentException("Not allowed to put key '" + ArcInvocationContext.KEY_INTERCEPTOR_BINDINGS);
+        }
+        getDelegateForWrite(m.size()).putAll(m);
+    }
+
+    @Override
+    public Object remove(Object key) {
+        if (ArcInvocationContext.KEY_INTERCEPTOR_BINDINGS.equals(key)) {
+            throw new IllegalArgumentException("Not allowed to remove key '" + ArcInvocationContext.KEY_INTERCEPTOR_BINDINGS);
+        } else if (delegate == null) {
+            return null;
+        } else {
+            return delegate.remove(key);
+        }
+    }
+
+    @Override
+    public int size() {
+        return getDelegateForRead().size() + 1;
+    }
+
+    @Override
+    public Collection values() {
+        throw new UnsupportedOperationException("Not allowed to get all values from the context data map");
+    }
+
+    @Override
+    public Object put(String key, Object value) {
+        if (ArcInvocationContext.KEY_INTERCEPTOR_BINDINGS.equals(key)) {
```

Comment:
I think since `KEY_INTERCEPTOR_BINDINGS` is an expectation of the particular Quarkus integration we shouldn't allow it to be replaced by some random 3rd party extension.